libsolidity/lsp/ReferenceCollector.cpp

```diff
@@ -0,0 +1,192 @@
+):
+	m_declaration{_declaration},
+	m_sourceIdentifierName{_sourceIdentifierName.empty() ? _declaration.name() : _sourceIdentifierName}
+{
+}
+
+std::vector<Reference> ReferenceCollector::collect(
+	Declaration const* _declaration,
+	ASTNode const& _ast,
+	std::string const& _sourceIdentifierName
+)
+{
+	if (!_declaration)
+		return {};
+
+	ReferenceCollector collector(*_declaration, _sourceIdentifierName);
+	_ast.accept(collector);
+	return std::move(collector.m_resultingReferences);
+}
+
+std::vector<Reference> ReferenceCollector::collect(
+	ASTNode const* _sourceNode,
+	SourceUnit const& _sourceUnit
+)
+{
+	if (!_sourceNode)
+		return {};
+
+	auto output = vector<Reference>{};
+
+	if (auto const* identifier = dynamic_cast<Identifier const*>(_sourceNode))
+	{
+		for (auto const* declaration: allAnnotatedDeclarations(identifier))
+			output += collect(declaration, _sourceUnit, identifier->name());
+	}
+	else if (auto const* identifierPath = dynamic_cast<IdentifierPath const*>(_sourceNode))
+	{
+		solAssert(identifierPath->path().size() >= 1, "");
+		output += collect(identifierPath->annotation().referencedDeclaration, _sourceUnit, identifierPath->path().back());
+	}
+	else if (auto const* memberAccess = dynamic_cast<MemberAccess const*>(_sourceNode))
+		output += collect(memberAccess->annotation().referencedDeclaration, _sourceUnit, memberAccess->memberName());
+	else if (auto const* declaration = dynamic_cast<Declaration const*>(_sourceNode))
+		output += collect(declaration, _sourceUnit, declaration->name());
+	else
+		lspAssert(false, ErrorCode::InternalError, "Unhandled AST node "s + typeid(*_sourceNode).name());
+
+	return output;
+}
+
+bool ReferenceCollector::visit(ImportDirective const& _import)
+{
+	if (_import.name() == m_sourceIdentifierName)
+		return true;
+	return false;
+}
+
+void ReferenceCollector::endVisit(ImportDirective const& _import)
+{
+	for (auto const& symbolAlias: _import.symbolAliases())
+	{
+		if (
+			m_sourceIdentifierName == *symbolAlias.alias &&
+			symbolAlias.symbol &&
+			symbolAlias.symbol->annotation().referencedDeclaration == &m_declaration
+		)
+		{
+			m_resultingReferences.emplace_back(symbolAlias.location, DocumentHighlightKind::Read);
+			break;
+		}
+		else if (m_sourceIdentifierName == *symbolAlias.alias)
+		{
+			m_resultingReferences.emplace_back(symbolAlias.location, DocumentHighlightKind::Text);
+			break;
+		}
+	}
+}
+
+bool ReferenceCollector::tryAddReference(Declaration const* _declaration, SourceLocation const& _location)
+{
+	if (&m_declaration != _declaration)
+		return false;
+
+	m_resultingReferences.emplace_back(_location, m_kind);
+	return true;
+}
+
+void ReferenceCollector::endVisit(Identifier const& _identifier)
+{
+	if (auto const* declaration = _identifier.annotation().referencedDeclaration)
+		tryAddReference(declaration, _identifier.location());
+
+	for (auto const* declaration: _identifier.annotation().candidateDeclarations + _identifier.annotation().overloadedDeclarations)
```

Comment:
The type is very short. Why not spell it out for more clarity?
```suggestion
	if (Declaration const* declaration = _identifier.annotation().referencedDeclaration)
		tryAddReference(declaration, _identifier.location());

	for (Declaration const* declaration: _identifier.annotation().candidateDeclarations + _identifier.annotation().overloadedDeclarations)
```

Reply:
Wasn't the name of the variable and the right hand side clear enough with `declaration` and `referencedDeclaration`, that we're talking about a declaration here? ;-)